src/core/string_map.cc

```diff
@@ -101,6 +101,74 @@ sds StringMap::Find(std::string_view key) {
   return GetValue(str);
 }
 
+void StringMap::RandomPair(sds* key, sds* value) {
+  auto itr = begin();
+  for (unsigned int i = 0; i < (rand() % Size()); ++i)
+    ++itr;
+  *key = itr->first;
+  if (value != nullptr)
+    *value = itr->second;
+  return;
+}
+
+unsigned int StringMap::RandomPairsUnique(unsigned int count, sds* keys, sds* vals) {
+  unsigned int total_size = Size();
+  unsigned int index = 0;
+  if (count > total_size)
+    count = total_size;
+
+  auto itr = begin();
+  unsigned int picked = 0, remaining = count;
+  while (picked < count && itr != end()) {
+    double random_double = ((double)rand()) / RAND_MAX;
+    double threshold = ((double)remaining) / (total_size - index);
+    if (random_double <= threshold) {
+      keys[picked] = itr->first;
+      if (vals) {
+        vals[picked] = itr->second;
+      }
+      remaining--;
+      picked++;
+    }
+    ++itr;
+    index++;
+  }
+  return picked;
+}
+
+void StringMap::RandomPairs(unsigned int count, sds* keys, sds* vals) {
+  typedef std::pair<unsigned int, unsigned int> RandomPick;
+  std::vector<RandomPick> picks;
+  unsigned int total_size = Size();
+
+  for (unsigned int i = 0; i < count; ++i) {
```

Comment:
you can use a `std::map<uint32_t, uint32_t>`.  The key represents the index, and value represents the number of times you encountered it (we need map here, to since it's an ordered container). Each time you loop here, you check if the index `rand() % total_size` already exists. If it does you increment it's occurrence (the value) by 1. Otherwise (if the index is not already inserted) you insert it with an associated value of `1` (since it's the first occurrence).  

That way you will:

1. Get rid of `std::sort`.
2. Simplify the code below, since now you don't need two while loops. You only need a for loop, and for each element, you output `occurrence` number of times

Reply:
Good suggestion, but I don't think it matters that much. Sort only gets slow for a really big number of elements and for it the i/o time is much larger

Reply:
No no, I do not care about the performance here, I don't expect that `std::sort` will have any measurable impact since it the count will always be relatively small. But by using a `map` it will simplify the implemetnaion of this function and make it more readable

Reply:
I prefer to keep as is. sort() is more implicit when std::map is used. Personally, the current version has better readabliy by making sort obvious.

Reply:
So a) I think stl associative ordered containers are used exactly for that, so I would argue that vector + sort is an antipattern. b) how is what you wrote simpler than this 3 lines? No pair accesses, no nested while loops no nothing:

```
for(auto it = begin(); it < end() && picks_sz < count ; ++it) {
  auto [key, frequency] = *it;
  keys.insert(keys.end(), frequency, key);
  picks_sz += frequency;
}
```

I might a have a mistake on the above code but that's the gist of it